src/packet.c

```diff
@@ -0,0 +1,72 @@
+#include "packet.h"
+
+
+Packet to_packet(struct sk_buff *skb) {
+    Packet packet = {};
+
+    if (!skb) {
+        return packet;
+    }
+
+    packet.ip_header = ip_hdr(skb);
+    if (packet.ip_header) {
+        switch (packet.ip_header->protocol)
+        {
+            case IPPROTO_TCP:
+                packet.layer4.tcp_header = tcp_hdr(skb);
+                break;
+            
+            default:
+                break;
+        }
+    }
+    return packet;
+}
+
+void modify_ip_header(Packet *packet, Rule *matched_rule) {
+    if (matched_rule->action.filter.src.ip) {
+        packet->ip_header->saddr = matched_rule->action.filter.src.ip;
+    }
+    if (matched_rule->action.filter.dst.ip) {
+        packet->ip_header->daddr = matched_rule->action.filter.dst.ip;
+    }
+    packet->ip_header->check = 0;
+    packet->ip_header->check = ip_fast_csum((unsigned char *)packet->ip_header, packet->ip_header->ihl);
+}
+
+void modify_tcp_header(Packet *packet, Rule *matched_rule) {
+    if (matched_rule->action.filter.src.port) {
+        packet->layer4.tcp_header->source = matched_rule->action.filter.src.port;
```

Comment:
Maybe validate packet->layer4.tcp_header is not null before accessing it.
You can do it once for all downstream usages.